### ISSUE TITLE
Analysis for new OBS multiple measures API

### ISSUE THREAD

**Issue by rafatower** [CartoDB/camshaft#297]
Implement an analysis with the following parameters:

```js
var PARAMS = {
    source: Node.PARAM.NODE(Node.GEOMETRY.POINT, Node.GEOMETRY.POLYGON),
    numerators: Node.PARAM.ARRAY(Node.PARAM.STRING()),
    denominators: Node.PARAM.ARRAY(Node.PARAM.STRING()),
    normalization: Node.PARAM.ARRAY(Node.PARAM.STRING()),
    geom_ids: Node.PARAM.ARRAY(Node.PARAM.STRING()),
    numerator_timespans: Node.PARAM.ARRAY(Node.PARAM.STRING()),
    column_names: Node.PARAM.ARRAY(Node.PARAM.STRING())
};
```

Note that elements of the params `denominators`, `geom_ids` and `numerator_timespans` can be `null`.

Here are [some examples](https://gist.github.com/rafatower/b24612af8a123f60bfdd891cad4b2897#full-example-with-normalization-timespans-and-boundaries) about how they shall be used to query `OBS_GetMeta` and `OBS_GetData` to obtain multiple measures in one request.

The work to be done can be based upon this branch by rochoa: https://github.com/CartoDB/camshaft/compare/data-observatory-multiple-measures

@nobuti can you please confirm that this interface between UI and analysis is fine? thanks!

**Comment by nobuti** [comment]
It looks great 💪

**Comment by rafatower** [comment]
I tested null values in metadata input for `OBS_GetMeta` and posted a question to get confirmation: https://github.com/CartoDB/observatory-extension/issues/288

Hopefully if we can use `null` values that way then the code can be kept simple.

**Comment by dgaubert** [comment]
Hey @rafatower! 

I have a quick question related to the params that we send to `OBS_GetMeta`. The following query is created by camshaft's node implemented in `data-observatory-multiple-measures` branch:

```
WITH _summary AS (
     SELECT ST_SetSRID(ST_Extent(the_geom), 4326) extent,
          count(*)::INT numgeoms,
          ARRAY_AGG(DISTINCT ST_GeometryType(the_geom)) AS geom_types,
          ARRAY_AGG((the_geom, cartodb_id)::geomval) geomvals,
          SUM(ST_Area(the_geom)) sumarea
     FROM (select * from atm_machines limit 2) AS _camshaft_do_measure_analysis
 ),
 _meta AS (
     SELECT cdb_dataservices_client.OBS_GetMeta(
         extent, ('[{"numer_id": "es.ine.t2_2", "target_area": ' || sumarea || '},{"numer_id": "es.ine.t2_1", "target_area": ' || sumarea || '}]')::JSON, 1, 1, numgeoms
     ) as meta
     FROM _summary
 ),
 _data AS (
   SELECT id AS __obs_id__, (data->0->>'value')::Numeric AS females, (data->0->>'value')::Numeric AS males
   FROM cdb_dataservices_client.OBS_GetData(
     (SELECT geomvals FROM _summary),
     (SELECT jsonb_set(meta::jsonb, '{0,normalization}',
         CASE
           WHEN meta->0->>'numer_aggregate' = 'sum' AND geom_types = '{ST_Point}'
             THEN '"area"'::jsonb
           ELSE '"prenormalized"'::jsonb
         END
     )::json FROM _meta, _summary)
   ))
 SELECT _camshaft_do_measure_analysis.*, _data.females, _data.males
 FROM _data, (select * from atm_machines limit 2) AS _camshaft_do_measure_analysis
 WHERE _camshaft_do_measure_analysis.cartodb_id = _data.__obs_id__
```
Just taking into account `_meta` CTE:

```
 _meta AS (
     SELECT cdb_dataservices_client.OBS_GetMeta(
         extent, ('[{"numer_id": "es.ine.t2_2", "target_area": ' || sumarea || '},{"numer_id": "es.ine.t2_1", "target_area": ' || sumarea || '}]')::JSON, 1, 1, numgeoms
     ) as meta
     FROM _summary
 )
```

What is the param `target_area`? and should I keep it?

Thanks in advance

**Comment by rafatower** [comment]
From the [docs](https://carto.com/docs/carto-engine/data/measures-functions#obsgetmetaextent-geometry-metadata-json-maxtimespanrank-maxscorerank-targetgeoms):

> Instead of aiming to have target_geoms in the area of the geometry passed as extent, fill this area. Unit is square degrees WGS84. Set this to 0 if you want to use the smallest source geometry for this element of metadata, for example if you’re passing in points.

By looking at the code it is passing `sumarea` (the sum of all areas in 4326). All that is used for area normalization AFAIK.

**Comment by rafatower** [comment]
@dgaubert an important remark: you don't need to keep the logic to choose the normalization type. Actually that is done by the analysis client (that already had to query the metadata discovery functions to populate all the ID's).

Sorry, I almost forgot to mention that.

**Comment by dgaubert** [comment]
I've made a first attempt with the implementation, with the following input:

```js
{
    type: 'data-observatory-multiple-measures',
    params: {
        source: {
            type: 'source',
            params: {
                query: 'select * from atm_machines limit 2'
            }
        },
        numerators: ['es.ine.t2_2', 'es.ine.t2_1'],
        normalizations: ['denominated', 'denominated'],
        denominators: ['es.ine.t1_1', 'es.ine.t1_1'],
        geom_ids: ['es.ine.the_geom', 'es.ine.the_geom'],
        numerator_timespans: ['2011', '2011'],
        column_names: ['females', 'males']
    }
}
```
camshaft builds the following query:
```sql
WITH _source AS (
 select * from atm_machines limit 2
),
_summary AS (
  SELECT
    ST_SetSRID(ST_Extent(the_geom), 4326) extent,
    count(*)::INT numgeoms
  FROM _source
),
_meta AS (
    SELECT cdb_dataservices_client.OBS_GetMeta(
        extent, ('[{"numer_id": "es.ine.t2_2","denom_id": "es.ine.t1_1","normalization": "denominated","geom_id": "es.ine.the_geom","numer_timespan": "2011"},{"numer_id": "es.ine.t2_1","denom_id": "es.ine.t1_1","normalization": "denominated", "geom_id": "es.ine.the_geom","numer_timespan": "2011"}]')::JSON, 1, 1, numgeoms
    ) as meta
    FROM _summary
),
_data AS (
  SELECT id AS __obs_id__, (data->0->>'value')::Numeric AS females, (data->0->>'value')::Numeric AS males
  FROM cdb_dataservices_client.OBS_GetData(
    (SELECT ARRAY_AGG((the_geom, cartodb_id)::geomval) FROM _source),
    (SELECT meta FROM _meta)
  ) AS _camshaft_do_measure_analysis_data
)
SELECT _source.*, _data.females, _data.males
FROM _data, _source
WHERE _source.cartodb_id = _data.__obs_id__
```
do you see anything weird? do you have tips? improvements?

Any help is more than welcome.

**Comment by dgaubert** [comment]
I'm worried about lasts params passed to `OBS_GetMeta`.

```sql
    SELECT cdb_dataservices_client.OBS_GetMeta(
        extent, ('[{"numer_id": "es.ine.t2_2","denom_id": "es.ine.t1_1","normalization": "denominated","geom_id": "es.ine.the_geom","numer_timespan": "2011"},{"numer_id": "es.ine.t2_1","denom_id": "es.ine.t1_1","normalization": "denominated", "geom_id": "es.ine.the_geom","numer_timespan": "2011"}]')::JSON, 1, 1, numgeoms
    ) as meta
    FROM _summary
```

Do we need them? In your examples I can't see them but in the previous version are there.

**Comment by dgaubert** [comment]
And the version with `null` values for non mandatory input params:

```sql
WITH _source AS (
 select * from atm_machines limit 2
),
_summary AS (
  SELECT
    ST_SetSRID(ST_Extent(the_geom), 4326) extent,
    count(*)::INT numgeoms
  FROM _source
),
_meta AS (
  SELECT
    cdb_dataservices_client.OBS_GetMeta(
      extent,
      ('[{"numer_id": "es.ine.t2_2","denom_id": null,"normalization": "denominated","geom_id": null,"numer_timespan": null},{"numer_id": "es.ine.t2_1","denom_id": null,"normalization": "denominated","geom_id": null,"numer_timespan": null}]'):
:JSON, 1, 1, numgeoms
    ) as meta
  FROM _summary
),
_data AS (
  SELECT id AS __obs_id__, (data->0->>'value')::Numeric AS females, (data->1->>'value')::Numeric AS males
  FROM cdb_dataservices_client.OBS_GetData(
    (SELECT ARRAY_AGG((the_geom, cartodb_id)::geomval) FROM _source),
    (SELECT meta FROM _meta)
  ) AS _camshaft_do_measure_analysis_data
)
SELECT _source.*, _data.females, _data.males
FROM _data, _source
WHERE _source.cartodb_id = _data.__obs_id__
```

**Comment by rafatower** [comment]
The queries looks ok to me. I'll try and check them in staging/prod

> I'm worried about lasts params passed to `OBS_GetMeta`.

From the [docs](https://carto.com/docs/carto-engine/data/measures-functions#obsgetmetaextent-geometry-metadata-json-maxtimespanrank-maxscorerank-targetgeoms) the last params are:
- `max_timespan_rank` set to 1. It defaults to 1
- `max_boundary_score_rank` set to 1 and defaults to 1 too.
- `target_geoms`: passes the `numgeoms` from the `_summary` subquery. Used to rank boundaries. In my example I just passed 1 geometry.

To sum up: I wouldn't be concerned at all, at least functionally.

I'll also take a look at the code.

Thanks!

**Comment by rafatower** [comment]
The only thing about the query in [this comment](https://github.com/CartoDB/camshaft/issues/297#issuecomment-302163850) is that `normalization` cannot be `denominated` if the denominator is set to `null`. But my understanding is that checking that is beyond the scope of camshaft

**Comment by rafatower** [comment]
PR here: https://github.com/CartoDB/camshaft/pull/298

**Comment by dgaubert** [comment]
Just for the record, I've tried the final query in staging an it worked as expected:

```sql
WITH _source AS (
  select * from ne_10m_populated_places_simple where name = 'Madrid'
),
_summary AS (
  SELECT
    ST_SetSRID(ST_Extent(the_geom), 4326) extent,
    count(*)::INT numgeoms
  FROM _source
),
_meta AS (
    SELECT cdb_dataservices_client.OBS_GetMeta(
        extent, ('[{"numer_id": "es.ine.t2_2","denom_id": "es.ine.t1_1","normalization": "denominated","geom_id": "es.ine.the_geom","numer_timespan": "2011"},{"numer_id": "es.ine.t2_1","denom_id": "es.ine.t1_1","normalization": "denominated", "geom_id": "es.ine.the_geom","numer_timespan": "2011"}]')::JSON, 1, 1, numgeoms
    ) as meta
    FROM _summary
),
_data AS (
  SELECT id AS __obs_id__, (data->0->>'value')::Numeric AS females, (data->1->>'value')::Numeric AS males
  FROM cdb_dataservices_client.OBS_GetData(
    (SELECT ARRAY_AGG((the_geom, cartodb_id)::geomval) FROM _source),
    (SELECT meta FROM _meta)
  ) AS _camshaft_do_measure_analysis_data
)
SELECT _source.*, _data.females, _data.males
FROM _data, _source
WHERE _source.cartodb_id = _data.__obs_id__
```

returns:

```sql
-[ RECORD 1 ]--------+---------------------------------------------------
cartodb_id           | 7265
the_geom             | 0101000020E61000007E1E71477D7B0DC0B65691D273334440
the_geom_webmercator | 0101000020110F0000AC48CBC8150A19C11B3BBA371CC95241
checkme              | 0
ls_match             | 1
rank_min             | 8
rank_max             | 13
pop_other            | 3673427
pop_min              | 50437
pop_max              | 5567000
namediff             | 0
megacity             | 1
diffascii            | 0
labelrank            | 3
natscale             | 300
scalerank            | 1
ls_name              | Madrid
meganame             | Madrid
diffnote             |
note                 |
iso_a2               | ES
adm1name             | Comunidad de Madrid
adm0_a3              | ESP
adm0name             | Spain
sov_a3               | ESP
sov0name             | Kingdom of Spain
capin                |
nameascii            | Madrid
namealt              |
namepar              |
name                 | Madrid
featurecla           | Admin-0 capital
geonameid            | 3675707
changed              | 0
longitude            | -3.683351686
latitude             | 40.4000262645
worldcity            | 1
capalt               | 0
adm0cap              | 1
updated_at           | 2014-10-19
created_at           | 2014-10-19
females              | 0.50649350649350649351
males                | 0.49350649350649350649
```

VAMOS!

**Comment by dgaubert** [comment]
I did a minor change to simplify definition of the analysis. @rafatower could take a quick look?

**Comment by rafatower** [comment]
LGTM

**Comment by jorgesancha** [comment]
Closing - ready in production
